src/main/java/tectech/thing/metaTileEntity/multi/godforge/util/GodforgeMath.java

```diff
@@ -293,8 +293,13 @@ public static boolean allowModuleConnection(MTEBaseModule module, MTEForgeOfGods
             return true;
         }
 
-        if (module instanceof MTEExoticModule && godforge.isUpgradeActive(QGPIU)) {
-            return true;
+        if (module instanceof MTEExoticModule) {
+            if (godforge.isUpgradeActive(QGPIU) && !((MTEExoticModule) module).isMagmatterModeOn()) {
+                return true;
+            }
+            if (godforge.isUpgradeActive(END) && ((MTEExoticModule) module).isMagmatterModeOn()) {
+                return true;
+            }
```

Comment:
Nitpick:
```suggestion
        if (module instanceof MTEExoticModule exoticizer) {
            if (godforge.isUpgradeActive(QGPIU) && !exoticizer.isMagmatterModeOn()) {
                return true;
            }
            if (godforge.isUpgradeActive(END) && exoticizer.isMagmatterModeOn()) {
                return true;
            }
```
Additionally, is it supposed to be an inverted check on the QGPIU upgrade?